File: 63-unique-paths-ii/unique-paths-ii.js
// Bottom-Up Dynamic Programming
// Time and Space: O(n * m)
var uniquePathsWithObstacles = function(obstacleGrid) {
    if (obstacleGrid[0][0] === 1) return 0
    const n = obstacleGrid.length, m = obstacleGrid[0].length

    const dp = [...new Array(n+1)].map(a => [...new Array(m + 1)].fill(0))
    dp[0][0] = 1

    for (let r = 0; r < n; r++) {
        for (let c = 0; c < m; c++) {
            if (obstacleGrid[r][c] === 1) continue;

            if (r-1 >= 0 && obstacleGrid[r-1][c] !== 1) { // can we go down
                dp[r][c] += dp[r-1][c]
            }

            if (c-1 >= 0 && obstacleGrid[r][c-1] !== 1) { // can we go right
                dp[r][c] += dp[r][c-1]
            }
        }
    }
    return dp[n-1][m-1]
};

File: 63-unique-paths-ii/unique-paths-ii.java
class Solution {
    // Bottom-Up Dynamic Programming
    // Time and Space: O(n * m)
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int n = obstacleGrid.length; int m = obstacleGrid[0].length;
        int dp[][] = new int[n+2][m+2];
        if (obstacleGrid[n-1][m-1] != 1) dp[n-1][m-1] = 1; // base case

        for (int r = n-1; r >= 0; r--) {
            for (int c = m-1; c >= 0; c--) {
                if (r == n-1 && c == m -1) continue;
                if (obstacleGrid[r][c] == 1) continue;
                int nx1 = dp[r+1][c]; int n2 = dp[r][c+1];
                int v = 0;
                if (r+1 < n && obstacleGrid[r+1][c] != 1) {
                    v+= nx1;
                }
                if (c+1 < m && obstacleGrid[r][c+1] != 1)  {
                    v += n2;
                }
                dp[r][c] = v;
            }
        }
        return dp[0][0];
    }
}